File: 110-balanced-binary-tree/balanced-binary-tree.java
class Solution {
    public boolean isBalanced(TreeNode root) {
        return dfs(root) != -1;
    }

    private int dfs(TreeNode node) {
        if (node == null) return 0; // base case: height is 0 for null node
        
        int leftHeight = dfs(node.left);
        if (leftHeight == -1) return -1; // early return if left subtree is unbalanced
        
        int rightHeight = dfs(node.right);
        if (rightHeight == -1) return -1; // early return if right subtree is unbalanced
        
        if (Math.abs(leftHeight - rightHeight) > 1) return -1; // current node is unbalanced

        return 1 + Math.max(leftHeight, rightHeight); // return height of the current subtree
    }
}

File: 110-balanced-binary-tree/balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
    const dfs = (node) => {
        if (!node) return 0;

        let leftHeight = dfs(node.left);
        let rightHeight = dfs(node.right);

        if (leftHeight === -1 || rightHeight === -1 || Math.abs(leftHeight - rightHeight) > 1) {
            return -1;
        }

        return Math.max(leftHeight, rightHeight) + 1;
    };

    return dfs(root) !== -1;
};
